Sources/MongoSwift/SDAM.swift

```diff
@@ -231,29 +254,111 @@ public struct TopologyDescription {
         }
     }
 
+    /// An internal function used to determine if there are suitable servers to read from in a replica set.
+    internal func hasSuitableReadServer(_ readPref: ReadPreference) -> Bool {
+        var suitableModes: [ServerType]
+
+        switch readPref.mode {
+        case ReadPreference.Mode.primary:
+            suitableModes = [ServerType.rsPrimary]
+        case ReadPreference.Mode.secondary:
+            suitableModes = [ServerType.rsSecondary]
+        default:
+            suitableModes = [ServerType.rsSecondary, ServerType.rsPrimary]
+        }
+
+        var candidates: [ServerDescription] = servers.filter { suitableModes.contains($0.type) }
+
+        // If read preference specifies a max staleness, first filter out too stale servers
+        if let maxStalenessSeconds = readPref.maxStalenessSeconds {
+            candidates = candidates.filter { (candidate: ServerDescription) in
+                let staleness = { () -> Double in
+                    let delta = { (serverDesc: ServerDescription) -> Double in
+                        (serverDesc.lastUpdateTime!.timeIntervalSinceReferenceDate -
```

Comment:
we're doing a lot of force unwrapping of optionals throughout this block. generally, we only want to do it if we are sure that the values will be present, because otherwise Swift crashes. 
are we sure that they will be present? does the spec define behavior for cases where they are not?

Reply:
`lastUpdateTime` is not necessarily present, so I've added protections for when it isn't. 
`lastWriteDate` should always be present in the `ismaster` for secondaries and primaries, and all candidates are filtered to be primaries or secondaries. If you want me to make this check more explicit, let me know.

Reply:
perfect - let's just add a comment somewhere around here explaining then that rationale.

Reply:
Done